packages/server/src/core/router.ts

```diff
@@ -104,27 +92,27 @@ export type inferHandlerInput<TProcedure extends Procedure<any>> =
 /**
  * @internal
  */
-type inferHandlerFn<TProcedures extends Record<string, Procedure<any>>> = <
+type inferHandlerFn<TProcedures extends ProcedureRecord> = <
   TProcedure extends TProcedures[TPath],
   TPath extends keyof TProcedures & string,
 >(
   path: TPath,
   ...args: inferHandlerInput<TProcedure>
 ) => Promise<inferProcedureOutput<TProcedure>>;
 
-type DecorateProcedure<TProcedure extends Procedure<any>> = (
+type DecorateProcedure<TProcedure extends AnyProcedure> = (
   input: ProcedureArgs<TProcedure['_def']>[0],
 ) => Promise<TProcedure['_def']['_output_out']>;
 
-type assertProcedure<T> = T extends Procedure<any> ? T : never;
-
 /**
  * @internal
  */
 type DecoratedProcedureRecord<TProcedures extends ProcedureRouterRecord> = {
   [TKey in keyof TProcedures]: TProcedures[TKey] extends AnyRouter
     ? DecoratedProcedureRecord<TProcedures[TKey]['_def']['record']>
-    : DecorateProcedure<assertProcedure<TProcedures[TKey]>>;
+    : TProcedures[TKey] extends AnyProcedure
```

Comment:
do you need this check here? TProcedures can only be AnyRouter or AnyProcedure so feels redundant, or is the TS compiler not satisfied otherwise?

Reply:
Unfortunately yes, I'm not sure why exactly TS is unable to narrow it down. Possibly because it's an indexed type.